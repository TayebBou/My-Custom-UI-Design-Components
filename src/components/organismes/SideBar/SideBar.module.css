.sidebar {
    position: fixed;
    height: 100vh;
    width: 450px;
    z-index: 1000;
    left: -450px;
    background-color: white;
    animation: appear 0.5s forwards;
}

.out {
    animation: disappear 0.5s forwards;
}

.close-button {
    background-color: transparent;
    border: 1px solid #6bb6f0;
    border-radius: 5px;
    width: 40px;
    height: 40px;
    display: flex;
}

.close-button:hover {
    transform: scale(1.1);
    transition: transform 0.3s;
}

.close-icon {
    align-self: center;
    margin: auto;
}

.sidebar-header {
    padding: 2em 60px 1em 40px;
    margin-bottom: 2em;
}

@keyframes appear {
    0%{
        transform: translateX(0%);
    }
    100%{
        transform: translateX(100%);
    }
}

@keyframes disappear {
    0%{
        transform: translateX(100%);
    }
    100%{
        transform: translateX(0%);
    }
}