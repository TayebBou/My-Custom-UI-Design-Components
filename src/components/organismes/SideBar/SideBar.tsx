import { FC, useState } from 'react'
import Button from '../../atoms/Button/Button'
import Icon from '../../atoms/Icon/Icon'
import styles from './SideBar.module.css'
import close from '../../../assets/images/cancel.png'

type SideBarProps = {
  children: React.ReactNode
  onClose: () => void
}

const SideBar: FC<SideBarProps> = (props) => {
  const { children, onClose } = props
  const [isClosing, setIsClosing] = useState(false)

  const handleClose = () => {
    setIsClosing(true)
    setTimeout(() => {
      onClose()
    }, 500)
  }

  return (
    <div className={styles.sidebar + ' ' + (isClosing ? styles.out : null)}>
      <div className={styles['sidebar-header']}>
        <Button className={styles['close-button']} onClick={handleClose}>
          <Icon className={styles['close-icon']} alt="close icon" src={close} size="24px" />
        </Button>
      </div>
      {children}
    </div>
  )
}

export default SideBar
