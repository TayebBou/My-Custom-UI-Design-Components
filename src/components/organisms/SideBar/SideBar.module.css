.sidebar {
  position: fixed;
  top: 0;
  height: 100vh;
  overflow-y: auto;
  width: 46%;
  z-index: 1000;
  left: -46%;
  background-color: white;
  animation: appear 0.5s forwards;
}

.out {
  animation: disappear 0.5s forwards;
}

.close-button {
  background-color: transparent;
  border: 1px solid #6bb6f0;
  border-radius: 5px;
  width: 44px;
  height: 44px;
  display: flex;
}

.close-button:hover {
  transform: scale(1.1);
  transition: transform 0.3s;
}

.close-icon {
  align-self: center;
  margin: auto;
}

.sidebar-header {
  padding: 2em 60px 1em 40px;
  margin-bottom: 2em;
}

@keyframes appear {
  0% {
    transform: translateX(0%);
  }
  100% {
    transform: translateX(100%);
  }
}

@keyframes disappear {
  0% {
    transform: translateX(100%);
  }
  100% {
    transform: translateX(0%);
  }
}

@media (max-width: 750px) {
  .sidebar {
    width: 50%;
    left: -50%;
  }
}

@media (max-width: 700px) {
  .sidebar {
    width: 390px;
    left: -390px;
  }
}

@media (max-width: 450px) {
  .sidebar {
    width: 90%;
    left: -90%;
  }
}

@media (max-width: 350px) {
  .sidebar {
    width: 95%;
    left: -95%;
  }
}
