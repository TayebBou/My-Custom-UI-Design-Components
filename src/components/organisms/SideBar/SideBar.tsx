import { FC, useCallback, useEffect, useRef, useState } from 'react'
import Button from '../../atoms/Button/Button'
import Icon from '../../atoms/Icon/Icon'
import styles from './SideBar.module.css'
import close from '../../../assets/images/cancel.png'

type SideBarProps = {
  children: React.ReactNode
  onClose: () => void
  closeSideBar?: boolean
}

const SideBar: FC<SideBarProps> = (props) => {
  const { children, onClose, closeSideBar } = props
  const [isClosing, setIsClosing] = useState(false)
  const [isFontClosing, setIsFontClosing] = useState(false)
  const [isSideBarOpen, setIsSideBarOpen] = useState(true)
  const timeout1 = useRef<NodeJS.Timeout | null>(null)
  const timeout2 = useRef<NodeJS.Timeout | null>(null)

  const handleClose = useCallback(() => {
    setIsFontClosing(true)
    setIsClosing(true)
    timeout1.current = setTimeout(() => {
      setIsSideBarOpen(false)
    }, 170)
    timeout2.current = setTimeout(() => {
      onClose()
    }, 500)
  }, [onClose])

  useEffect(() => {
    if (closeSideBar) {
      handleClose()
    }
    return () => {
      if(timeout1.current !== null) {
        clearTimeout(timeout1.current)
      }
      if(timeout2.current !== null) {
        clearTimeout(timeout2.current)
      }
    }
  }, [closeSideBar, handleClose])

  return (
    <>
      <div className={styles.sidebar + ' ' + (isClosing ? styles.out : '')}>
        <div className={styles['sidebar-header']}>
          <Button className={styles['close-button']} onClick={handleClose}>
            <Icon
              className={styles['close-icon']}
              alt="close icon"
              src={close}
              size="26px"
            />
          </Button>
        </div>
        {children}
      </div>
      {isSideBarOpen && (
        <div className={styles['black-font'] + ' ' + (isFontClosing ? styles['out-black'] : '')} onClick={handleClose} />
      )}
    </>
  )
}

SideBar.defaultProps = {
  closeSideBar: false,
}

export default SideBar
