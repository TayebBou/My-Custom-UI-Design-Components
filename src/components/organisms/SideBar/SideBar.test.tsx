// @ts-nocheck

import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import SideBar from "./SideBar";

const onClose = jest.fn();

describe("SideBar component", () => {
  beforeEach(() => {
    onClose.mockClear();
  });
  test("should render correctly with props", () => {
    // Arrange
    const { container } = render(
      <SideBar onClose={onClose} closeSideBar={true}>
        children
      </SideBar>
    );
    // Assert
    expect(container).toMatchSnapshot();
  });
  test("should render correctly without props", () => {
    // Arrange
    const { container } = render(<SideBar />);
    // Assert
    expect(container).toMatchSnapshot();
  });
  test("when user click on close button the onClose function is invoked", async () => {
    // Arrange
    render(<SideBar onClose={onClose}>children</SideBar>);
    const closeButton = screen.getByRole("button");
    // Act
    fireEvent.click(closeButton);
    // Assert
    // the onClose function is executed after 0.5s ( time for cancel animation )
    await waitFor(() => {
      expect(onClose).toHaveBeenCalled();
    });
  });
  // to close the sideBar dynamically we use the closeSideBar props
  test("when user render sideBar with closeSideBar prop at true the onClose function is invoked", async () => {
    // Arrange
    render(
      <SideBar onClose={onClose} closeSideBar={true}>
        children
      </SideBar>
    );
    // Assert
    // the onClose function is executed after 0.5s ( time for cancel animation )
    await waitFor(() => {
      expect(onClose).toHaveBeenCalled();
    });
  });
});
