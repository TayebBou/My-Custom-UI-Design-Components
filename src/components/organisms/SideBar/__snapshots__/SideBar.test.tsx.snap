// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar component should render correctly with props 1`] = `
<div>
  <div
    class="sidebar out"
  >
    <div
      class="sidebar-header"
    >
      <button
        class="button close-button"
        type="button"
      >
        <img
          alt="close icon"
          class="close-icon"
          src="cancel.png"
          style="width: 26px; height: 26px; filter: none;"
        />
      </button>
    </div>
    children
  </div>
  <div
    class="black-font out-black"
  />
</div>
`;

exports[`SideBar component should render correctly without props 1`] = `
<div>
  <div
    class="sidebar "
  >
    <div
      class="sidebar-header"
    >
      <button
        class="button close-button"
        type="button"
      >
        <img
          alt="close icon"
          class="close-icon"
          src="cancel.png"
          style="width: 26px; height: 26px; filter: none;"
        />
      </button>
    </div>
  </div>
  <div
    class="black-font "
  />
</div>
`;
